hbase-server/src/main/java/org/apache/hadoop/hbase/wal/WALKeyImpl.java

```diff
@@ -195,6 +195,42 @@ public WALKeyImpl(final byte[] encodedRegionName,
         mvcc, null, null);
   }
 
+  /**
+   * Copy constructor that takes in an existing WALKeyImpl plus some extended attributes.
+   * Intended for coprocessors to add annotations to a system-generated WALKey
+   * for persistence to the WAL.
+   * @param key
+   * @param extendedAttributes
+   */
+  public WALKeyImpl(WALKeyImpl key,
```

Comment:
Instead of CP creating a new WALKey object, we can give API to add the ext attributes? It allows to add any custom key value pair so that later CPs or Replication EPs an make use.  So we might not even allow the CP to return a brand new WALKey object. (?)

Reply:
Also WALKeyImpl is marked LimitedPrivate({HBaseInterfaceAudience.REPLICATION}) I can see. So what this patch trying to do is to expose the class for CPs also.  I can not see where WALKeyImpl to be used by Replication area. WALKey was enough.   Anyways for this CP hook, IMO its enough to pass WALKey interface.  We can add the API to add ext attributes into the interface. That is just adding ext attributes. It should be fine as it is not exposing any setters which takes other key attributes of the object.